File: variant-v/java-service/src/main/java/com/flashsale/config/RedisManager.java
package com.flashsale.config;

import org.redisson.api.RedissonClient;

import java.security.MessageDigest;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class RedisManager {
    private final RedissonClient primaryClient;
    private final List<String> nodes;
    private final Map<Integer, RedissonClient> nodeClients = new ConcurrentHashMap<>();
    
    public RedisManager(RedissonClient primaryClient, List<String> nodes) {
        this.primaryClient = primaryClient;
        this.nodes = nodes;
    }
    
    public RedissonClient getClientForSku(String skuId) {
        int nodeIndex = getNodeIndexForSku(skuId);
        return getClientForNode(nodeIndex);
    }
    
    private RedissonClient getClientForNode(int nodeIndex) {
        return nodeClients.computeIfAbsent(nodeIndex, idx -> primaryClient);
    }
    
    public int getNodeIndexForSku(String skuId) {
        try {
            MessageDigest md = MessageDigest.getInstance("MD5");
            byte[] hash = md.digest(skuId.getBytes());
            int hashValue = 0;
            for (int i = 0; i < 4; i++) {
                hashValue = (hashValue << 8) | (hash[i] & 0xFF);
            }
            return Math.abs(hashValue) % nodes.size();
        } catch (Exception e) {
            return 0;
        }
    }
    
    public RedissonClient getPrimaryClient() {
        return primaryClient;
    }
}

File: variant-v/csharp-service/Services/RedisManager.cs
using StackExchange.Redis;
using System.Security.Cryptography;
using System.Text;

namespace FlashSale.Api.V.Services;

public class RedisManager
{
    private readonly ConnectionMultiplexer[] _connections;
    private readonly IDatabase[] _databases;
    
    public RedisManager(string[] nodes)
    {
        _connections = new ConnectionMultiplexer[nodes.Length];
        _databases = new IDatabase[nodes.Length];
        
        for (int i = 0; i < nodes.Length; i++)
        {
            var config = ConfigurationOptions.Parse(nodes[i]);
            config.ConnectTimeout = 5000;
            config.SyncTimeout = 5000;
            config.AbortOnConnectFail = false;
            
            _connections[i] = ConnectionMultiplexer.Connect(config);
            _databases[i] = _connections[i].GetDatabase();
        }
    }
    
    public IDatabase GetDatabaseForSku(string skuId)
    {
        int nodeIndex = GetNodeIndexForSku(skuId);
        return _databases[nodeIndex];
    }
    
    public ConnectionMultiplexer GetConnectionForSku(string skuId)
    {
        int nodeIndex = GetNodeIndexForSku(skuId);
        return _connections[nodeIndex];
    }
    
    public IDatabase[] GetAllDatabases() => _databases;
    
    private int GetNodeIndexForSku(string skuId)
    {
        using (var md5 = MD5.Create())
        {
            var hash = md5.ComputeHash(Encoding.UTF8.GetBytes(skuId));
            int hashValue = BitConverter.ToInt32(hash, 0);
            return Math.Abs(hashValue) % _databases.Length;
        }
    }
}
